### DISCUSSION TITLE
Plural translation issue

### DISCUSSION THREAD

**Discussion by sonix-github** [koreader/koreader#9719] [Q&A]
Hello,

I am translating to Slovak language KOReader and by verification I have found an issue with plural translations. If I translate the strings via Weblate page, the translated strings are incorrectly shown in KOReader. It does not show correct "plurals" for count 0,5,6,7.... It seems like a bug, but I am not sure.

Could someone advise with this issue?

Thanks

![Weblate_Plural_Translation_Highlight_2](https://user-images.githubusercontent.com/29783895/199008782-24684a32-9798-4bb0-8e3c-2dd54a593aee.jpg)

![Plural_Translation_Issue_Highlight_2](https://user-images.githubusercontent.com/29783895/199008809-f3564002-41ee-44fa-808d-ade782253189.jpg)

**Comment by Frenzie** [comment]
At a glance it looks okay to me. Where's the screenshot from exactly?
![image](https://user-images.githubusercontent.com/202757/199029883-bdd18763-de03-4ab9-be02-8a6d13831384.png)

**Reply by Frenzie** [reply]
E.g., if it's from your local emulator you might need to run something like `git submodule update` or `make pot` to sync or update your translations to the latest version.

**Reply by sonix-github** [reply]
The screenshot is taken from pocketbook 626 and it is from "History" list. I am using nightly build "koreader-pocketbook-v2022.10-50-g4d8e2f0e_2022-10-29". Your screenshot looks correct. It is strange. If I switch to Czech language it has also incorrect words with plurals. I will delete whole koreader folder and copy new one.

**Reply by Frenzie** [reply]
The switch is followed by a restart, correct? I realize it sounds like a stupid question but just making sure. Just this Wednesday I had one of those "I could've sworn this setting was correct so I stupidly didn't double check it for a few minutes while doing advanced troubleshooting" moments. ;-)

**Reply by sonix-github** [reply]
Yes, I switch the language and then restart KOReader. Now I am on new copy of koreader extracted from nightly build "koreader-pocketbook-v2022.10-50-g4d8e2f0e_2022-10-29" and the plurals are still incorrect for count 0,5,6,7.... I switched to English + restart and then back to SLovak + restart and still it is incorrect.

1. Could it be something only around pocketbook?
2. Could you please clarify me where is located "logic" for choosing the correct plural string from msgstr[X] or how can I debug it?

Thanks

This is probably the part where the string is taken for translation from Slovak koreader.po
```
#: plugins/coverbrowser.koplugin/listmenu.lua:416
msgid "Finished – 1 page"
msgid_plural "Finished – %1 pages"
msgstr[0] "Dokončené – 1 strana"
msgstr[1] "Dokončené – %1 strany"
msgstr[2] "Dokončené – %1 strany"
msgstr[3] "Dokončených – %1 strán"
```

**Reply by Frenzie** [reply]
The logic can be seen on your screenshot:

```
"Plural-Forms: nplurals=4; plural=(n % 1 == 0 && n == 1 ? 0 : n % 1 == 0 && n "
">= 2 && n <= 4 ? 1 : n % 1 != 0 ? 2: 3);\n"
```

You can follow it along yourself or just stuff in a number in a C-like language.
```
(n % 1 == 0 && n == 1 ? 0 : n % 1 == 0 && n >= 2 && n <= 4 ? 1 : n % 1 != 0 ? 2: 3)
```

For example, in the JavaScript console in your browser:
```
n=0
0
(n % 1 == 0 && n == 1 ? 0 : n % 1 == 0 && n >= 2 && n <= 4 ? 1 : n % 1 != 0 ? 2: 3)
3
n=10
10
(n % 1 == 0 && n == 1 ? 0 : n % 1 == 0 && n >= 2 && n <= 4 ? 1 : n % 1 != 0 ? 2: 3)
3
n=1
1
(n % 1 == 0 && n == 1 ? 0 : n % 1 == 0 && n >= 2 && n <= 4 ? 1 : n % 1 != 0 ? 2: 3)
0 
```

I wrote a little Lua conversion script [here](https://github.com/koreader/koreader/blob/94d3d3b4872ee6767bb40a1b9372eb2d950bf4f8/frontend/gettext.lua#L96-L137) which spits out an equivalent Lua function.

In this case that's:
```
function return function(n) if n % 1 == 0 and n == 1  then return 0  elseif  n % 1 == 0 and n >= 2 and n <= 4  then return 1  elseif  n % 1 ~= 0  then return 2 else return  3 end end
```

The outcome is identical.

**Reply by sonix-github** [reply]
Hmm, the formula is functioning as expected on computer. But I do not understand why if the n=3, the msgstr[3] string is not returned. I have enabled in developer options debug logging + verbose debug logging, unfortunately I do not see debug lines (in crash.log) from gettext.lua (`logger.dbg("gettext: plural function", plural_func_str)`).
Could you please advise how can I make visible debug lines from gettext.lua?

**Reply by Frenzie** [reply]
My bad, I was a little over a week out of date with my translations and as it turns out I suggested to do something that shouldn't have been done. As a result the metadata (i.e., the plurals formula) wasn't parsed by the program meaning all languages with more complicated plurals were treated like English. Thanks for spotting it!

I think it should work in tomorrow's nightly, but please let me know if it doesn't.

**Reply by Frenzie** [reply]
See https://github.com/koreader/koreader/pull/9726

I'm surprised it didn't create an auto-link.

**Reply by sonix-github** [reply]
I have downloaded / copied last nightly build "koreader-pocketbook-v2022.10-75-g4a355f78_2022-11-02" and after restart the plurals are now OK. By the way I still do not see debug lines (in crash.log) from gettext.lua `logger.dbg("gettext: plural function", plural_func_str)`
